### PR TITLE
[AF-1792]  Migration of space information from system.git to config.git + HA improvements

### PR DESCRIPTION
**Pull request by adrielparedes**
# AF-1792

## Description

Several things happen in this PR that can't be split in several commits:

### Migration from system.git to config.git

All the information about spaces and repositories used to live in system.git project. That project makes that every change that was need to be done locks repository for all users. 
Now, all the information about each space is stored inside config.git in every space. So lock will happen only for that space and not for the whole file system. Also migration tool was modified to include previous changes.

### Logical Deletion

When you delete a Space or Repository, they are marked to be deleted. User will continue working as it was used to. Nothing from user perspectiva but they way that deleted spaces are shown. 
* A deleted Repository disappear from Library View and is marked to be deleted with a flag in config.git
* A deleted Space doesn't disappear from Spaces View but users can't get into that Space. In the UI the space is marked to be deleted. Also is marked to be deleted with a flag in config.git.

### Physical Deletion

Every one minute a FileSystemDeleteWorker runs to delete all the repositories and spaces that are marked to be deleted. This runs in all nodes, but only one will be able to execute those operations.

### HA Improvements

Many events where added to improve HA experience. Also visual improvements to show correct assets/spaces/repositories/contributors counters and deleted spaces right identification.

### RHSSO

RHSSO configuration was included to let BC consume contributors from SSO List.

## Collaborators

* Paulo Rego
* Pere Fernandez Pere
* Eder Ignatowicz
* Adriel Paredes

## PRs related

* https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/970
* https://github.com/kiegroup/appformer/pull/715
* https://github.com/kiegroup/kie-wb-common/pull/2692
* https://github.com/kiegroup/jbpm-wb/pull/1362
* https://github.com/kiegroup/kie-wb-distributions/pull/936

### CODE REVIEW

**Comment by adrielparedes** [comment]
jenkins retest this

**Comment by adrielparedes** [comment]
jenkins retest this

**Comment by ederign** [comment]
@tomasdavidorg @jakubschwan a suggestion to test this is to import all the samples in more than one space in 7.4, run the migration tool and check if everything works in Adriel's new structure.

File: kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-project-migration/src/main/java/org/kie/workbench/common/project/cli/util/ConfigGroupToSpaceInfoConverter.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        if (defaultGroupId == null || defaultGroupId.trim().isEmpty()) {
+            defaultGroupId = getSanitizedDefaultGroupId(extractName(groupConfig));
+        }
+
+        return defaultGroupId;
+    }
+
+    private String getSanitizedDefaultGroupId(final String proposedGroupId) {
+        //Only [A-Za-z0-9_\-.] are valid so strip everything else out
+        return proposedGroupId != null ? proposedGroupId.replaceAll("[^A-Za-z0-9_\\-.]",
```
Comment (pefernan) [inline kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-project-migration/src/main/java/org/kie/workbench/common/project/cli/util/ConfigGroupToSpaceInfoConverter.java]: Maybe keeping the REGEX in a constant... but that's only a personal preference, so keep it as a suggestion.

File: kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-project-migration/src/main/java/org/kie/workbench/common/project/cli/util/ConfigGroupToSpaceInfoConverter.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+        boolean oldConfigGroup = false;
+
+        final String oldOwner = configGroup.getConfigItemValue("owner");
+        if (oldOwner != null) {
+            oldConfigGroup = true;
+            contributors.add(new Contributor(oldOwner,
+                                             ContributorType.OWNER));
+        }
+
+        ConfigItem<List<String>> oldContributors = configGroup.getConfigItem("contributors");
```
Comment (pefernan) [inline kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-project-migration/src/main/java/org/kie/workbench/common/project/cli/util/ConfigGroupToSpaceInfoConverter.java]: Same with "contributors"

File: kie-wb-common-screens/kie-wb-common-examples-screen/kie-wb-common-examples-screen-backend/src/main/java/org/kie/workbench/common/screens/examples/backend/server/BaseProjectImportService.java [hunk 3]
```diff
@@ -171,23 +167,20 @@ protected String readDescription(final Module module) {
...
-        }
+        RepositoryConfiguration configuration = new RepositoryConfiguration(env);
+
+        configuration.add(EnvironmentParameters.AVOID_INDEX,
+                          true);
 
-        repositoryConfig.addConfigItem(configurationFactory.newConfigItem(EnvironmentParameters.AVOID_INDEX,
-                                                                          "true"));
+        configuration.add(EnvironmentParameters.SPACE,
+                          "system");
```
Comment (pefernan) [inline kie-wb-common-screens/kie-wb-common-examples-screen/kie-wb-common-examples-screen-backend/src/main/java/org/kie/workbench/common/screens/examples/backend/server/BaseProjectImportService.java]: Constant!

File: kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-system-to-space-configuration-migration/src/main/java/org/kie/workbench/common/system/space/configuration/SystemToSpaceConfigurationMigrationTool.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2018 Red Hat, Inc. and/or its affiliates.
```
Comment (tomasdavidorg) [inline kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-system-to-space-configuration-migration/src/main/java/org/kie/workbench/common/system/space/configuration/SystemToSpaceConfigurationMigrationTool.java]:
```suggestion
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
```

File: kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-system-to-space-configuration-migration/src/main/resources/META-INF/services/org.kie.workbench.common.migration.cli.MigrationTool [hunk 5]
```diff
@@ -0,0 +1,17 @@
+#
+# Copyright 2018 Red Hat, Inc. and/or its affiliates.
```
Comment (tomasdavidorg) [inline kie-wb-common-cli/kie-wb-common-cli-tools/kie-wb-common-cli-system-to-space-configuration-migration/src/main/resources/META-INF/services/org.kie.workbench.common.migration.cli.MigrationTool]:
```suggestion
# Copyright 2019 Red Hat, Inc. and/or its affiliates.
```

**Comment by tomasdavidorg** [comment]
So far I have found one minor issue. If you move mouse to any space, dialog `This space is marked for removal...` is always displayed.
Screenshot from 2019-06-14 16-26-27

**Comment by tomasdavidorg** [comment]
@adrielparedes If you stay in Spaces screen for some time, will the deleted spaces disappear or you have to swich to another screen?

**Comment by tomasdavidorg** [comment]
@adrielparedes @pefernan @paulovmr I tried the migration tool and it seems that it does not work properly. I run the migration tool on niogit with two spaces. The first space with one project and the second space with two projects. After I copied the migrated niogit to the workbench I saw only the spaces without projects. 

I also tried the `test_niogit.zip`. I ran the migration tool and copied the migrated niogit tool to the clean workbench. I cannot see the projects inside the MySpace.

I think the config structure (SpaceInfo.json) of the space was not correctly migrated:
`{"^EncodedType":"org.guvnor.structure.organizationalunit.config.SpaceInfo","^ObjectID":"1","name":"MySpace","deleted":false,"defaultGroupId":"com.myspace","contributors":{"^EncodedType":"java.util.ArrayList","^ObjectID":"2","^Value":[{"^EncodedType":"org.guvnor.structure.contributors.Contributor","^ObjectID":"3","username":"admin","type":{"^EncodedType":"org.guvnor.structure.contributors.ContributorType","^EnumStringValue":"OWNER"}}]},"repositories":{"^EncodedType":"java.util.ArrayList","^ObjectID":"4","^Value":[]},"securityGroups":{"^EncodedType":"java.util.ArrayList","^ObjectID":"5","^Value":[]}}`

Does it work for you?

**Comment by pefernan** [comment]
@tomasdavidorg could you try this zip (https://github.com/kiegroup/kie-wb-common/blob/master/kie-wb-common-cli/kie-wb-common-cli-migration-tool/src/test/resources/org.kie.workbench.common.migration/test_64x_niogit.zip) instead?

**Comment by paulovmr** [comment]
@tomasdavidorg I've fixed the "space is marked for removal" message in all spaces. Now it should only appear when the mouse goes over the yellow warning icon of a marked for removal space.

**Comment by pefernan** [comment]
jenkins retest this please

**Comment by pefernan** [comment]
@adrielparedes @tomasdavidorg pushed the migration tool issue fix.

@adrielparedes sorry for pushing directly

**Comment by pefernan** [comment]
@tomasdavidorg if you don't want to wait here you have a migration tool: https://mega.nz/#!EBxlgK6I!28K7t4p9sL0hFyFc3pNoswbDc6qoMOJ8wbkSAba7zw8

**Pull request merged**